lib/spack/spack/solver/asp.py

```diff
@@ -446,9 +448,18 @@ def pkg_version_rules(self, pkg):
             #    c) Numeric or string comparison
             v)
 
-        most_to_least_preferred = sorted(
-            self.possible_versions[pkg.name], key=keyfn, reverse=True
-        )
+        # Compute which versions appear only in packages.yaml
+        from_packages_yaml = self.versions_in_packages_yaml[pkg.name]
```

Comment:
Does this include versions set from

```
packages:
  foo:
    versions: [...]
```

If not, then I think we should rename the variables to be `*from_externals` instead of `*from_packages_yaml`.

Reply:
No, it doesn't account for version preferences - only for versions of external specs that are unknown to Spack. I'll change the name accordingly.

Reply:
Changed, rebased and pushed.